yjit/src/yjit.rs

```diff
@@ -46,11 +46,62 @@ pub fn yjit_enabled_p() -> bool {
     YJIT_ENABLED.load(Ordering::Acquire)
 }
 
+/// Make the call threshold available to C
+#[no_mangle]
+pub extern "C" fn rb_yjit_call_threshold() -> raw::c_ulong {
+    get_option!(call_threshold) as raw::c_ulong
+}
+
+// Counter to serve as a proxy for execution time, total number of calls
+static mut TOTAL_ENTRY_HITS: u64 = 0;
+
+// Number of calls used to estimate how hot an ISEQ is
+static CALL_COUNT_INTERV: u64 = 20;
+
 /// Test whether we are ready to compile an ISEQ or not
 #[no_mangle]
-pub extern "C" fn rb_yjit_threshold_hit(_iseq: IseqPtr, total_calls: u64) -> bool {
+pub extern "C" fn rb_yjit_threshold_hit(iseq: IseqPtr, total_calls: u64) -> bool {
+
     let call_threshold = get_option!(call_threshold) as u64;
-    return total_calls == call_threshold;
+
+    unsafe { TOTAL_ENTRY_HITS += 1; }
+
+    // Exit early if the threshold is not reached
+    if total_calls < call_threshold {
+        return false;
+    }
+
+    // We expect threshold 1 to compile everything immediately
+    // We need to compare for equality here because otherwise we
+    // can get a deaclock with ractors
+    if call_threshold < CALL_COUNT_INTERV {
+        return total_calls == call_threshold;
+    }
+
+    // Record the number of calls at the beginning of the interval
+    if total_calls == call_threshold - CALL_COUNT_INTERV {
```

Comment:
I'm confused. When `call_threshold` is 30 and `CALL_COUNT_INTERV` is 20, this is hit when `total_calls` is 10. However, `total_calls < call_threshold` is true in that case, so it would have already returned above at "Exit early if the threshold is not reached" (`if total_calls < call_threshold {`). So it seems like this expression is never true. What am I missing here?

Reply:
Whoops! 😅